File: sdks/js/src/crypto/__tests__/session.test.ts
// End-to-end session tests: two independent "devices" (Alice, Bob) each with
// their own identity, running the real X3DH handshake and Double Ratchet
// exchange against each other exactly as two real clients would. This is the
// most important test file in the whole chat-crypto layer — it's what
// stands in for cross-implementation test vectors, verifying the two ends
// of the algorithm actually agree, not just that each function looks right
// in isolation.
import {
  generateIdentityKeyPair,
  generateSignedPrekey,
  generateOneTimePrekeys,
} from '../identity';
import { initiateX3DH, receiveX3DH, buildPrekeyMessageHeader, type PrekeyBundle } from '../x3dh';
import {
  initRatchetAsSender,
  initRatchetAsReceiver,
  ratchetEncrypt,
  ratchetDecrypt,
  type RatchetState,
  type RatchetMessage,
} from '../doubleRatchet';
import { utf8ToBytes, bytesToUtf8, toBase64, fromBase64 } from '../bytes';

/** Simulates a device: its identity, one signed prekey, and a small
 * one-time-prekey pool, plus the bundle it would publish to the server. */
function makeDevice() {
  const identity = generateIdentityKeyPair();
  const signedPrekey = generateSignedPrekey(identity, 1);
  const oneTimePrekeys = generateOneTimePrekeys(1, 5);
  return { identity, signedPrekey, oneTimePrekeys };
}

function publishBundle(device: ReturnType<typeof makeDevice>, consumeOneTime: boolean): PrekeyBundle {
  // Peek, don't remove: the test later looks up this exact keyId on the
  // device's own pool to get its private half (as the device itself would
  // still hold it locally) — the server-side "this key is now consumed"
  // bookkeeping isn't what these tests exercise.
  const otp = consumeOneTime ? device.oneTimePrekeys[0] : undefined;
  return {
    deviceId: 'device',
    identityKey: toBase64(concatIdentity(device)),
    signedPrekeyId: device.signedPrekey.keyId,
    signedPrekey: toBase64(device.signedPrekey.keyPair.publicKey),
    signedPrekeySig: toBase64(device.signedPrekey.signature),
    oneTimePrekeyId: otp?.keyId,
    oneTimePrekey: otp ? toBase64(otp.keyPair.publicKey) : undefined,
  };
}

function concatIdentity(device: ReturnType<typeof makeDevice>): Uint8Array {
  const dh = device.identity.dh.publicKey;
  const sign = device.identity.sign.publicKey;
  const out = new Uint8Array(dh.length + sign.length);
  out.set(dh, 0);
  out.set(sign, dh.length);
  return out;
}

/** Runs the full handshake: Alice initiates against Bob's bundle, Bob
 * completes it from the resulting prekey message header. Returns both
 * sides' initialized ratchet state. */
async function establishSession(alice: ReturnType<typeof makeDevice>, bob: ReturnType<typeof makeDevice>, withOneTimePrekey: boolean) {
  const bundle = publishBundle(bob, withOneTimePrekey);
  const x3dh = await initiateX3DH(alice.identity, bundle);
  const header = buildPrekeyMessageHeader(alice.identity, x3dh);

  const aliceState = await initRatchetAsSender(
    x3dh.sharedSecret,
    alice.identity.dh.publicKey,
    x3dh.theirIdentityDHPublicKey,
    bob.signedPrekey.keyPair.publicKey
  );

  // Bob looks up the one-time prekey by id (if the header names one) before
  // it's consumed — this is what the real device keeps locally by keyId.
  const usedOtp = header.oneTimePrekeyId != null
    ? bob.oneTimePrekeys.find((k) => k.keyId === header.oneTimePrekeyId) ?? null
    : null;

  const bobSharedSecret = await receiveX3DH(
    bob.identity,
    bob.signedPrekey.keyPair,
    usedOtp ? usedOtp.keyPair : null,
    alice.identity.dh.publicKey,
    fromBase64(header.ephemeralKey)
  );

  const bobState = initRatchetAsReceiver(
    bobSharedSecret,
    bob.identity.dh.publicKey,
    alice.identity.dh.publicKey,
    bob.signedPrekey.keyPair
  );

  return { aliceState, bobState, header };
}

describe('X3DH handshake', () => {
  it('both sides derive the same shared secret (with a one-time prekey)', async () => {
    const alice = makeDevice();
    const bob = makeDevice();
    const bundle = publishBundle(bob, true);
    const x3dh = await initiateX3DH(alice.identity, bundle);
    const header = buildPrekeyMessageHeader(alice.identity, x3dh);
    const usedOtp = bob.oneTimePrekeys.find((k) => k.keyId === header.oneTimePrekeyId)!;

    const bobSecret = await receiveX3DH(
      bob.identity,
      bob.signedPrekey.keyPair,
      usedOtp.keyPair,
      alice.identity.dh.publicKey,
      x3dh.ephemeralPublicKey
    );

    expect(toBase64(bobSecret)).toBe(toBase64(x3dh.sharedSecret));
  });

  it('both sides derive the same shared secret without a one-time prekey', async () => {
    const alice = makeDevice();
    const bob = makeDevice();
    const bundle = publishBundle(bob, false); // pool not consumed -> bundle has none
    const x3dh = await initiateX3DH(alice.identity, bundle);

    const bobSecret = await receiveX3DH(
      bob.identity,
      bob.signedPrekey.keyPair,
      null,
      alice.identity.dh.publicKey,
      x3dh.ephemeralPublicKey
    );

    expect(toBase64(bobSecret)).toBe(toBase64(x3dh.sharedSecret));
  });

  it('rejects a bundle with an invalid signed-prekey signature', async () => {
    const alice = makeDevice();
    const bob = makeDevice();
    const bundle = publishBundle(bob, false);
    bundle.signedPrekeySig = toBase64(new Uint8Array(64)); // garbage signature
    await expect(initiateX3DH(alice.identity, bundle)).rejects.toThrow(/signature/);
  });
});

describe('Double Ratchet session', () => {
  it('Alice can send and Bob can decrypt the first message', async () => {
    const alice = makeDevice();
    const bob = makeDevice();
    const { aliceState, bobState } = await establishSession(alice, bob, true);

    const msg = await ratchetEncrypt(aliceState, utf8ToBytes('hello bob'));
    const plaintext = await ratchetDecrypt(bobState, msg);
    expect(bytesToUtf8(plaintext)).toBe('hello bob');
  });

  it('supports a full back-and-forth conversation', async () => {
    const alice = makeDevice();
    const bob = makeDevice();
    const { aliceState, bobState } = await establishSession(alice, bob, true);

    const m1 = await ratchetEncrypt(aliceState, utf8ToBytes('hi bob'));
    expect(bytesToUtf8(await ratchetDecrypt(bobState, m1))).toBe('hi bob');

    const m2 = await ratchetEncrypt(bobState, utf8ToBytes('hi alice'));
    expect(bytesToUtf8(await ratchetDecrypt(aliceState, m2))).toBe('hi alice');

    const m3 = await ratchetEncrypt(aliceState, utf8ToBytes('how are you'));
    expect(bytesToUtf8(await ratchetDecrypt(bobState, m3))).toBe('how are you');

    const m4 = await ratchetEncrypt(bobState, utf8ToBytes('good, you?'));
    expect(bytesToUtf8(await ratchetDecrypt(aliceState, m4))).toBe('good, you?');
  });

  it('handles several messages within the same sending chain', async () => {
    const alice = makeDevice();
    const bob = makeDevice();
    const { aliceState, bobState } = await establishSession(alice, bob, true);

    for (let i = 0; i < 5; i++) {
      const msg = await ratchetEncrypt(aliceState, utf8ToBytes(`message ${i}`));
      expect(bytesToUtf8(await ratchetDecrypt(bobState, msg))).toBe(`message ${i}`);
    }
  });

  it('decrypts out-of-order messages within a chain via the skipped-key cache', async () => {
    const alice = makeDevice();
    const bob = makeDevice();
    const { aliceState, bobState } = await establishSession(alice, bob, true);

    const m0 = await ratchetEncrypt(aliceState, utf8ToBytes('zero'));
    const m1 = await ratchetEncrypt(aliceState, utf8ToBytes('one'));
    const m2 = await ratchetEncrypt(aliceState, utf8ToBytes('two'));

    // Bob receives them out of order: 2, 0, 1.
    expect(bytesToUtf8(await ratchetDecrypt(bobState, m2))).toBe('two');
    expect(bytesToUtf8(await ratchetDecrypt(bobState, m0))).toBe('zero');
    expect(bytesToUtf8(await ratchetDecrypt(bobState, m1))).toBe('one');
  });

  it('decrypts messages out of order across a ratchet direction change', async () => {
    const alice = makeDevice();
    const bob = makeDevice();
    const { aliceState, bobState } = await establishSession(alice, bob, true);

    // Alice sends two messages in her first chain.
    const a1 = await ratchetEncrypt(aliceState, utf8ToBytes('a1'));
    const a2 = await ratchetEncrypt(aliceState, utf8ToBytes('a2'));
    // Bob only receives a1 for now (a2 arrives late, tested below).
    expect(bytesToUtf8(await ratchetDecrypt(bobState, a1))).toBe('a1');

    // Bob replies — this is a NEW ratchet key for Bob, so when Alice
    // receives it she'll need to DH-ratchet.
    const b1 = await ratchetEncrypt(bobState, utf8ToBytes('b1'));
    expect(bytesToUtf8(await ratchetDecrypt(aliceState, b1))).toBe('b1');

    // Alice replies again (new chain, since she just ratcheted on receipt of b1).
    const a3 = await ratchetEncrypt(aliceState, utf8ToBytes('a3'));
    expect(bytesToUtf8(await ratchetDecrypt(bobState, a3))).toBe('a3');

    // The late a2 (from Alice's OLD chain, before the ratchet) must still
    // decrypt via Bob's skipped-key cache, recorded when he ratcheted on a3.
    expect(bytesToUtf8(await ratchetDecrypt(bobState, a2))).toBe('a2');
  });

  it('fails closed on a tampered ciphertext', async () => {
    const alice = makeDevice();
    const bob = makeDevice();
    const { aliceState, bobState } = await establishSession(alice, bob, true);

    const msg = await ratchetEncrypt(aliceState, utf8ToBytes('secret'));
    const tampered: RatchetMessage = { header: msg.header, ciphertext: msg.ciphertext.slice(0, -4) + 'AAAA' };
    await expect(ratchetDecrypt(bobState, tampered)).rejects.toThrow();
  });

  it('a third party without the session keys cannot decrypt', async () => {
    const alice = makeDevice();
    const bob = makeDevice();
    const eve = makeDevice();
    const { aliceState } = await establishSession(alice, bob, true);
    // Eve establishes her own (unrelated) session with Bob and tries to
    // decrypt Alice's message using her own state — simulates an outsider
    // who intercepted the ciphertext but has no valid session for it.
    const { bobState: eveViewOfBob } = await establishSession(eve, bob, true);

    const msg = await ratchetEncrypt(aliceState, utf8ToBytes('for bob only'));
    await expect(ratchetDecrypt(eveViewOfBob, msg)).rejects.toThrow();
  });
});

File: sdks/dart/test/crypto/session_test.dart
// NOTE: written without a Dart toolchain available to run it — mirrors
// sdks/js/src/crypto/__tests__/session.test.ts, which IS verified (25
// passing tests covering X3DH agreement, bidirectional messaging, in-order
// and out-of-order delivery across ratchet direction changes, tamper
// detection, and third-party isolation). Run `melos test` (or `dart test`
// from sdks/dart/) before trusting this port.
//
// Two independent "devices" (Alice, Bob) each with their own identity,
// running the real X3DH handshake and Double Ratchet exchange against each
// other exactly as two real clients would — this is what stands in for
// cross-implementation test vectors here, verifying the two ends of the
// algorithm actually agree, not just that each function looks right alone.
import 'dart:typed_data';

import 'package:test/test.dart';
import 'package:applad/src/crypto/crypto.dart';

class _Device {
  final IdentityKeyPair identity;
  final SignedPrekey signedPrekey;
  final List<OneTimePrekey> oneTimePrekeys;
  _Device({required this.identity, required this.signedPrekey, required this.oneTimePrekeys});
}

Future<_Device> _makeDevice() async {
  final identity = await generateIdentityKeyPair();
  final signedPrekey = await generateSignedPrekey(identity, 1);
  final oneTimePrekeys = await generateOneTimePrekeys(1, 5);
  return _Device(identity: identity, signedPrekey: signedPrekey, oneTimePrekeys: oneTimePrekeys);
}

/// Peek, don't remove: the test later looks up this exact keyId on the
/// device's own pool to get its private half (as the device itself would
/// still hold it locally) — the server-side "this key is now consumed"
/// bookkeeping isn't what these tests exercise.
PrekeyBundle _publishBundle(_Device device, bool consumeOneTime) {
  final otp = consumeOneTime && device.oneTimePrekeys.isNotEmpty ? device.oneTimePrekeys.first : null;
  return PrekeyBundle(
    deviceId: 'device',
    identityKey: encodeIdentityPublicKey(device.identity),
    signedPrekeyId: device.signedPrekey.keyId,
    signedPrekey: toBase64(device.signedPrekey.keyPair.publicKey),
    signedPrekeySig: toBase64(device.signedPrekey.signature),
    oneTimePrekeyId: otp?.keyId,
    oneTimePrekey: otp != null ? toBase64(otp.keyPair.publicKey) : null,
  );
}

class _Session {
  final RatchetState aliceState;
  final RatchetState bobState;
  final PrekeyMessageHeader header;
  _Session({required this.aliceState, required this.bobState, required this.header});
}

Future<_Session> _establishSession(_Device alice, _Device bob, bool withOneTimePrekey) async {
  final bundle = _publishBundle(bob, withOneTimePrekey);
  final x3dh = await initiateX3DH(alice.identity, bundle);
  final header = buildPrekeyMessageHeader(alice.identity, x3dh);

  final aliceState = await initRatchetAsSender(
    x3dh.sharedSecret,
    alice.identity.dh.publicKey,
    x3dh.theirIdentityDHPublicKey,
    bob.signedPrekey.keyPair.publicKey,
  );

  OneTimePrekey? usedOtp;
  if (header.oneTimePrekeyId != null) {
    for (final k in bob.oneTimePrekeys) {
      if (k.keyId == header.oneTimePrekeyId) {
        usedOtp = k;
        break;
      }
    }
  }

  final bobSharedSecret = await receiveX3DH(
    bob.identity,
    bob.signedPrekey.keyPair,
    usedOtp?.keyPair,
    alice.identity.dh.publicKey,
    fromBase64(header.ephemeralKey),
  );

  final bobState = initRatchetAsReceiver(
    bobSharedSecret,
    bob.identity.dh.publicKey,
    alice.identity.dh.publicKey,
    bob.signedPrekey.keyPair,
  );

  return _Session(aliceState: aliceState, bobState: bobState, header: header);
}

void main() {
  group('X3DH handshake', () {
    test('both sides derive the same shared secret (with a one-time prekey)', () async {
      final alice = await _makeDevice();
      final bob = await _makeDevice();
      final bundle = _publishBundle(bob, true);
      final x3dh = await initiateX3DH(alice.identity, bundle);
      final header = buildPrekeyMessageHeader(alice.identity, x3dh);
      final usedOtp = bob.oneTimePrekeys.firstWhere((k) => k.keyId == header.oneTimePrekeyId);

      final bobSecret = await receiveX3DH(
        bob.identity,
        bob.signedPrekey.keyPair,
        usedOtp.keyPair,
        alice.identity.dh.publicKey,
        x3dh.ephemeralPublicKey,
      );

      expect(toBase64(bobSecret), equals(toBase64(x3dh.sharedSecret)));
    });

    test('both sides derive the same shared secret without a one-time prekey', () async {
      final alice = await _makeDevice();
      final bob = await _makeDevice();
      final bundle = _publishBundle(bob, false);
      final x3dh = await initiateX3DH(alice.identity, bundle);

      final bobSecret = await receiveX3DH(
        bob.identity,
        bob.signedPrekey.keyPair,
        null,
        alice.identity.dh.publicKey,
        x3dh.ephemeralPublicKey,
      );

      expect(toBase64(bobSecret), equals(toBase64(x3dh.sharedSecret)));
    });

    test('rejects a bundle with an invalid signed-prekey signature', () async {
      final alice = await _makeDevice();
      final bob = await _makeDevice();
      final bundle = _publishBundle(bob, false);
      final tampered = PrekeyBundle(
        deviceId: bundle.deviceId,
        identityKey: bundle.identityKey,
        signedPrekeyId: bundle.signedPrekeyId,
        signedPrekey: bundle.signedPrekey,
        signedPrekeySig: toBase64(Uint8List(64)), // garbage signature
      );
      expect(() => initiateX3DH(alice.identity, tampered), throwsA(anything));
    });
  });

  group('Double Ratchet session', () {
    test('Alice can send and Bob can decrypt the first message', () async {
      final alice = await _makeDevice();
      final bob = await _makeDevice();
      final session = await _establishSession(alice, bob, true);

      final msg = await ratchetEncrypt(session.aliceState, utf8ToBytes('hello bob'));
      final plaintext = await ratchetDecrypt(session.bobState, msg);
      expect(bytesToUtf8(plaintext), equals('hello bob'));
    });

    test('supports a full back-and-forth conversation', () async {
      final alice = await _makeDevice();
      final bob = await _makeDevice();
      final session = await _establishSession(alice, bob, true);

      final m1 = await ratchetEncrypt(session.aliceState, utf8ToBytes('hi bob'));
      expect(bytesToUtf8(await ratchetDecrypt(session.bobState, m1)), equals('hi bob'));

      final m2 = await ratchetEncrypt(session.bobState, utf8ToBytes('hi alice'));
      expect(bytesToUtf8(await ratchetDecrypt(session.aliceState, m2)), equals('hi alice'));

      final m3 = await ratchetEncrypt(session.aliceState, utf8ToBytes('how are you'));
      expect(bytesToUtf8(await ratchetDecrypt(session.bobState, m3)), equals('how are you'));

      final m4 = await ratchetEncrypt(session.bobState, utf8ToBytes('good, you?'));
      expect(bytesToUtf8(await ratchetDecrypt(session.aliceState, m4)), equals('good, you?'));
    });

    test('handles several messages within the same sending chain', () async {
      final alice = await _makeDevice();
      final bob = await _makeDevice();
      final session = await _establishSession(alice, bob, true);

      for (var i = 0; i < 5; i++) {
        final msg = await ratchetEncrypt(session.aliceState, utf8ToBytes('message $i'));
        expect(bytesToUtf8(await ratchetDecrypt(session.bobState, msg)), equals('message $i'));
      }
    });

    test('decrypts out-of-order messages within a chain via the skipped-key cache', () async {
      final alice = await _makeDevice();
      final bob = await _makeDevice();
      final session = await _establishSession(alice, bob, true);

      final m0 = await ratchetEncrypt(session.aliceState, utf8ToBytes('zero'));
      final m1 = await ratchetEncrypt(session.aliceState, utf8ToBytes('one'));
      final m2 = await ratchetEncrypt(session.aliceState, utf8ToBytes('two'));

      expect(bytesToUtf8(await ratchetDecrypt(session.bobState, m2)), equals('two'));
      expect(bytesToUtf8(await ratchetDecrypt(session.bobState, m0)), equals('zero'));
      expect(bytesToUtf8(await ratchetDecrypt(session.bobState, m1)), equals('one'));
    });

    test('decrypts messages out of order across a ratchet direction change', () async {
      final alice = await _makeDevice();
      final bob = await _makeDevice();
      final session = await _establishSession(alice, bob, true);

      final a1 = await ratchetEncrypt(session.aliceState, utf8ToBytes('a1'));
      final a2 = await ratchetEncrypt(session.aliceState, utf8ToBytes('a2'));
      expect(bytesToUtf8(await ratchetDecrypt(session.bobState, a1)), equals('a1'));

      final b1 = await ratchetEncrypt(session.bobState, utf8ToBytes('b1'));
      expect(bytesToUtf8(await ratchetDecrypt(session.aliceState, b1)), equals('b1'));

      final a3 = await ratchetEncrypt(session.aliceState, utf8ToBytes('a3'));
      expect(bytesToUtf8(await ratchetDecrypt(session.bobState, a3)), equals('a3'));

      // The late a2 (from Alice's OLD chain, before the ratchet) must still
      // decrypt via Bob's skipped-key cache, recorded when he ratcheted on a3.
      expect(bytesToUtf8(await ratchetDecrypt(session.bobState, a2)), equals('a2'));
    });

    test('fails closed on a tampered ciphertext', () async {
      final alice = await _makeDevice();
      final bob = await _makeDevice();
      final session = await _establishSession(alice, bob, true);

      final msg = await ratchetEncrypt(session.aliceState, utf8ToBytes('secret'));
      final tampered = RatchetMessage(
        header: msg.header,
        ciphertext: '${msg.ciphertext.substring(0, msg.ciphertext.length - 4)}AAAA',
      );
      expect(() => ratchetDecrypt(session.bobState, tampered), throwsA(anything));
    });

    test('a third party without the session keys cannot decrypt', () async {
      final alice = await _makeDevice();
      final bob = await _makeDevice();
      final eve = await _makeDevice();
      final session = await _establishSession(alice, bob, true);
      final eveSession = await _establishSession(eve, bob, true);

      final msg = await ratchetEncrypt(session.aliceState, utf8ToBytes('for bob only'));
      expect(() => ratchetDecrypt(eveSession.bobState, msg), throwsA(anything));
    });
  });
}
